owlkettle/guidsl.nim

```diff
@@ -193,42 +192,33 @@ proc gen(node: Node, stmts, parent: NimNode) =
       else:
         stmts.add(name)
     of NodeAttribute:
-      stmts.add(new_assignment(
-        new_dot_expr(parent, "has_" & node.name),
-        bind_sym("true")
-      ))
-      stmts.add(new_assignment(
-        new_dot_expr(parent, "val_" & node.name),
-        node.value
-      ))
+      stmts.add:
+        new_assignment(new_dot_expr(parent, "has_" & node.name), bind_sym("true"))
+      stmts.add:
+        new_assignment(new_dot_expr(parent, "val_" & node.name), node.value)
     of NodeEvent:
-      let typ = new_tree(nnkProcTy, [
-        node.callback.params, new_empty_node()
-      ])
+      let typ = new_tree(nnkProcTy, node.callback.params, new_empty_node())
       node.callback.name = new_empty_node()
-      stmts.add(new_assignment(
-        new_dot_expr(parent, node.event),
-        new_tree(nnkObjConstr, [
-          bindsym("Event").new_bracket_expr(typ),
-          new_tree(nnkExprColonExpr, [
-            ident("callback"), node.callback
-          ])
-        ])
-      ))
+
+      stmts.add:
+        gen_ast(parent, typ, node_event = ident node.event, node_callback = node.callback):
+          parent.node_event = Event[typ](callback: node_callback)
+
     of NodeBlock:
       for child in node.children:
         child.gen(stmts, parent)
     of NodeFor:
       var body = new_stmt_list()
       node.children[0].gen(body, parent)
-      stmts.add(new_tree(nnkForStmt, node.vars & @[
-        node.iter,
-        new_stmt_list(
-          new_call(bind_sym("custom_capture"),
-            node.vars.find_variables() & @[body]
-          )
-        )
-      ]))
+      stmts.add:
+        nnkForStmt.newTree:
```

Comment:
In this case the regular call (not callblock syntax) also is more readable to me, but I guess that is somewhat subjective.